csrc/optimization/optimize_consecutive_cast.cpp

```diff
@@ -0,0 +1,108 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir_utils.h>
+#include <optimization/opt_pass.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+class ConsecutiveCastPass : OptimizationPass {
+ public:
+  static void runPass(Fusion* fusion) {
+    auto is_cast_op = [](Expr* expr) {
+      if (expr->isA<UnaryOp>()) {
+        auto op = expr->as<UnaryOp>();
+        if (op->getUnaryOpType() == UnaryOpType::Cast) {
+          return true;
+        }
+      }
+      return false;
+    };
+
+    // NOTE: not the most efficient pass
+    std::unordered_map<Val*, Val*> replacement_map;
+    for (auto expr : fusion->exprs()) {
+      if (is_cast_op(expr)) {
+        bool mutated = false;
+        while (true) {
+          // in the loop, we just repetitively skip consecutive casts.
+          auto intermediate_cast = expr->input(0);
+          auto prev_expr = intermediate_cast->definition();
+          if (prev_expr != nullptr && is_cast_op(prev_expr)) {
+            auto original_dtype = prev_expr->input(0)->getDataType().value();
+            auto intermediate_dtype = intermediate_cast->getDataType().value();
+            auto out_dtype = expr->output(0)->getDataType().value();
+            // cases where skipping the intermediate cast is relatively safe,
+            // two conditions:
+            //   1. original_dtype is the same as out_dtype; or
+            //   2. we support direct cast from original_dtype to out_dtype.
+            // and
+            //   1. intermediate_dtype is the same type category as with
+            //   out_dtype; or
+            //   2. intermediate_dtype is a floating point while output is
+            //   integral;
+            if ((original_dtype == out_dtype ||
```

Comment:
This if condition is difficult to follow.

Reply:
Does the comment not help?
I can try to write more comment on top :laughing: